server.go

```diff
@@ -61,36 +74,142 @@ func Serve(ctx context.Context, c *websocket.Conn, execer Execer) error {
 
 		switch header.Type {
 		case proto.TypeStart:
+			if process != nil {
+				return errors.New("command already started")
+			}
+
 			var header proto.ClientStartHeader
 			err = json.Unmarshal(byt, &header)
 			if err != nil {
 				return xerrors.Errorf("unmarshal start header: %w", err)
 			}
-			process, err = execer.Start(ctx, mapToClientCmd(header.Command))
-			if err != nil {
-				return err
-			}
 
-			_ = sendPID(ctx, process.Pid(), wsNetConn)
-
-			var outputgroup errgroup.Group
-			outputgroup.Go(func() error {
-				return copyWithHeader(process.Stdout(), wsNetConn, proto.Header{Type: proto.TypeStdout})
-			})
-			outputgroup.Go(func() error {
-				return copyWithHeader(process.Stderr(), wsNetConn, proto.Header{Type: proto.TypeStderr})
-			})
-
-			go func() {
-				defer wsNetConn.Close()
-				_ = outputgroup.Wait()
-				err = process.Wait()
-				if exitErr, ok := err.(ExitError); ok {
-					_ = sendExitCode(ctx, exitErr.Code, wsNetConn)
-					return
+			command := mapToClientCmd(header.Command)
+
+			// Only allow TTYs with IDs to be reconnected.
+			if command.TTY && header.ID != "" {
+				// Enforce a consistent format for IDs.
+				_, err := uuid.Parse(header.ID)
+				if err != nil {
+					flog.Error("%s is not a valid uuid: %w", header.ID, err)
+				}
+
+				// Get an existing process or create a new one.
+				var rprocess *reconnectingProcess
+				rawRProcess, ok := reconnectingProcesses.Load(header.ID)
+				if ok {
+					rprocess, ok = rawRProcess.(*reconnectingProcess)
+					if !ok {
+						flog.Error("found invalid type in reconnecting process map for ID %s", header.ID)
+					}
+					process = rprocess.process
+				} else {
+					process, err = execer.Start(ctx, command)
+					if err != nil {
+						return err
+					}
+
+					rprocess = &reconnectingProcess{
+						activeConns: make(map[string]net.Conn),
+						process:     process,
+						timeout:     time.NewTimer(options.ReconnectingProcessTimeout),
+						// Default to buffer 1MB.
+						ringBuffer: ringbuffer.New(1 << 20),
+					}
+					reconnectingProcesses.Store(header.ID, rprocess)
+					go func() {
+						// Close if the inactive timeout occurs, or the context ends.
+						select {
+						case <-rprocess.timeout.C:
+							flog.Info("killing reconnecting process %s due to inactivity", header.ID)
+						case <-pctx.Done():
+						}
+						rprocess.Close()
+					}()
+					go func() {
+						buffer := make([]byte, 32*1024)
+						for {
+							// TODO: This is erroring with "read /dev/ptmx: input/output
+							// error" when the web socket closes causing reconnection to stop
+							// working.
+							read, err := rprocess.process.Stdout().Read(buffer)
```

Comment:
This errors when the original web socket closes and breaks future reconnects as well as any other concurrent connections.

Reply:
I think I have it sorted, seems the process is getting killed by the context which sucks because I forgot I already ran into that once before lol